File: skills/agents-orchestrator/scripts/compat_env.py
"""Canonical environment names with strict Agent Swarm compatibility.

The Runtime still has integration points that read ``AGENT_SWARM_*`` directly
(notably the persisted SQLite runtime and older entrypoints).  New callers use
``AGENTS_ORCHESTRATOR_*``.  This module keeps resolution in one place:

* canonical values are primary;
* a non-empty legacy value is an accepted fallback;
* conflicting dual definitions and partial identities fail closed;
* child/process boundaries can deliberately export both families.

The default runtime home and SQLite filename remain the existing values until
``state_store`` can consume this helper directly.  Keeping those defaults here
makes that later integration explicit and avoids silently stranding persisted
Runs during the package rename.
"""

from contextlib import contextmanager
import os
import pathlib


CANONICAL_PREFIX = "AGENTS_ORCHESTRATOR_"
LEGACY_PREFIX = "AGENT_SWARM_"
RUNTIME_HOME_DIRECTORY = ".agent-swarm"
RUNTIME_SQLITE_FILENAME = "runtime.sqlite3"

IDENTITY_SUFFIXES = (
    "ROOT_ID",
    "TASK_ID",
    "ATTEMPT_ID",
    "ACTOR_TOKEN",
)
BOUNDARY_SUFFIXES = IDENTITY_SUFFIXES + (
    "HOME",
    "SKILL_DIR",
)
TRANSIENT_IDENTITY_SUFFIXES = (
    "AGENT_ID",
    "EXECUTION_NONCE",
)


def canonical_name(suffix):
    return CANONICAL_PREFIX + suffix


def legacy_name(suffix):
    return LEGACY_PREFIX + suffix


def _nonempty(environment, name):
    value = environment.get(name)
    if isinstance(value, str):
        value = value.strip()
    return None if value is None or value == "" else value


def value(suffix, environment=None, default=None):
    """Resolve canonical > legacy while rejecting unequal dual definitions."""
    environment = os.environ if environment is None else environment
    primary_name = canonical_name(suffix)
    fallback_name = legacy_name(suffix)
    primary = _nonempty(environment, primary_name)
    fallback = _nonempty(environment, fallback_name)
    if primary is not None and fallback is not None and primary != fallback:
        raise ValueError(
            "conflicting orchestration environment: %s does not match %s"
            % (primary_name, fallback_name)
        )
    return primary if primary is not None else fallback if fallback is not None else default


def validate_identity(environment=None):
    """Reject a partially populated identity family, even if the other is full."""
    environment = os.environ if environment is None else environment
    for prefix in (CANONICAL_PREFIX, LEGACY_PREFIX):
        present = [
            suffix
            for suffix in IDENTITY_SUFFIXES
            if _nonempty(environment, prefix + suffix) is not None
        ]
        if present and len(present) != len(IDENTITY_SUFFIXES):
            missing = [
                prefix + suffix
                for suffix in IDENTITY_SUFFIXES
                if suffix not in present
            ]
            raise ValueError(
                "partial orchestration identity: missing %s" % ", ".join(missing)
            )
    return {
        suffix: value(suffix, environment)
        for suffix in IDENTITY_SUFFIXES
    }


def promote_canonical_environment(environment=None):
    """Make canonical-only process values visible to legacy integration code.

    Legacy-only environments are intentionally not mirrored as an import side
    effect.  Boundary exporters perform symmetric dual-family export.
    """
    environment = os.environ if environment is None else environment
    validate_identity(environment)
    suffixes = set(BOUNDARY_SUFFIXES)
    suffixes.update(
        name[len(CANONICAL_PREFIX) :]
        for name in tuple(environment)
        if name.startswith(CANONICAL_PREFIX)
        and len(name) > len(CANONICAL_PREFIX)
    )
    for suffix in sorted(suffixes):
        primary = _nonempty(environment, canonical_name(suffix))
        fallback = _nonempty(environment, legacy_name(suffix))
        if primary is not None and fallback is not None and primary != fallback:
            value(suffix, environment)  # raises the stable conflict diagnostic
        if primary is not None and fallback is None:
            environment[legacy_name(suffix)] = primary
    return environment


def export_both(values, *, base=None, scrub_identity=False):
    """Return an environment fragment with canonical and legacy names.

    ``values`` uses suffix keys such as ``ROOT_ID`` and ``HOME``.  When
    ``scrub_identity`` is true, stale identity/transient values are removed
    from a copied base environment before the supplied binding is exported.
    """
    result = dict(base or {})
    if scrub_identity:
        for suffix in IDENTITY_SUFFIXES + TRANSIENT_IDENTITY_SUFFIXES:
            result.pop(canonical_name(suffix), None)
            result.pop(legacy_name(suffix), None)
    for suffix, raw in values.items():
        if raw is None:
            continue
        rendered = str(raw)
        result[canonical_name(suffix)] = rendered
        result[legacy_name(suffix)] = rendered
    return result


@contextmanager
def process_boundary(values):
    """Temporarily install an exact dual-family child identity in ``os.environ``."""
    names = [
        prefix + suffix
        for prefix in (CANONICAL_PREFIX, LEGACY_PREFIX)
        for suffix in IDENTITY_SUFFIXES + TRANSIENT_IDENTITY_SUFFIXES
    ]
    names.extend(
        prefix + suffix
        for prefix in (CANONICAL_PREFIX, LEGACY_PREFIX)
        for suffix in ("HOME", "SKILL_DIR")
    )
    before = {name: os.environ.get(name) for name in names}
    try:
        for name in names:
            os.environ.pop(name, None)
        os.environ.update(export_both(values))
        yield
    finally:
        for name in names:
            os.environ.pop(name, None)
        for name, prior in before.items():
            if prior is not None:
                os.environ[name] = prior


def runtime_home(environment=None):
    """Resolve the compatibility runtime home used by the current SQLite layer."""
    environment = os.environ if environment is None else environment
    configured = value("HOME", environment)
    if configured:
        return pathlib.Path(configured).expanduser().resolve()
    return (pathlib.Path.home() / RUNTIME_HOME_DIRECTORY).resolve()


def runtime_sqlite_path(environment=None):
    """Return the current compatibility SQLite path for future store integration."""
    return runtime_home(environment) / RUNTIME_SQLITE_FILENAME


# Names intentionally mirror state_store's current integration surface.
def runtime_root(environment=None):
    return runtime_home(environment)


def db_path(environment=None):
    return runtime_sqlite_path(environment)

File: skills/agents-orchestrator/scripts/compat_env.ts
import { homedir } from "node:os";
import { resolve } from "node:path";

import { ValueError } from "./runtime_types.ts";

export const CANONICAL_PREFIX = "AGENTS_ORCHESTRATOR_";
export const LEGACY_PREFIX = "AGENT_SWARM_";
export const RUNTIME_HOME_DIRECTORY = ".agent-swarm";
export const RUNTIME_SQLITE_FILENAME = "runtime.sqlite3";

export const IDENTITY_SUFFIXES = ["ROOT_ID", "TASK_ID", "ATTEMPT_ID", "ACTOR_TOKEN"] as const;
export const BOUNDARY_SUFFIXES = [...IDENTITY_SUFFIXES, "HOME", "SKILL_DIR"] as const;
export const TRANSIENT_IDENTITY_SUFFIXES = ["AGENT_ID", "EXECUTION_NONCE"] as const;

export type Environment = Record<string, string | undefined>;

export function canonicalName(suffix: string): string {
  return `${CANONICAL_PREFIX}${suffix}`;
}

export function legacyName(suffix: string): string {
  return `${LEGACY_PREFIX}${suffix}`;
}

function nonempty(environment: Environment, name: string): string | undefined {
  const raw = environment[name];
  if (typeof raw !== "string") return undefined;
  const rendered = raw.trim();
  return rendered === "" ? undefined : rendered;
}

export function value(
  suffix: string,
  environment: Environment = process.env,
  defaultValue?: string,
): string | undefined {
  const primaryName = canonicalName(suffix);
  const fallbackName = legacyName(suffix);
  const primary = nonempty(environment, primaryName);
  const fallback = nonempty(environment, fallbackName);
  if (primary !== undefined && fallback !== undefined && primary !== fallback) {
    throw new ValueError(`conflicting orchestration environment: ${primaryName} does not match ${fallbackName}`);
  }
  return primary ?? fallback ?? defaultValue;
}

export function validateIdentity(environment: Environment = process.env): Record<string, string | undefined> {
  for (const prefix of [CANONICAL_PREFIX, LEGACY_PREFIX]) {
    const present = IDENTITY_SUFFIXES.filter((suffix) => nonempty(environment, `${prefix}${suffix}`) !== undefined);
    if (present.length > 0 && present.length !== IDENTITY_SUFFIXES.length) {
      const missing = IDENTITY_SUFFIXES
        .filter((suffix) => !present.includes(suffix))
        .map((suffix) => `${prefix}${suffix}`);
      throw new ValueError(`partial orchestration identity: missing ${missing.join(", ")}`);
    }
  }
  return Object.fromEntries(IDENTITY_SUFFIXES.map((suffix) => [suffix, value(suffix, environment)]));
}

export function promoteCanonicalEnvironment(environment: Environment = process.env): Environment {
  validateIdentity(environment);
  const suffixes = new Set<string>(BOUNDARY_SUFFIXES);
  for (const name of Object.keys(environment)) {
    if (name.startsWith(CANONICAL_PREFIX) && name.length > CANONICAL_PREFIX.length) {
      suffixes.add(name.slice(CANONICAL_PREFIX.length));
    }
  }
  for (const suffix of [...suffixes].sort()) {
    const primary = nonempty(environment, canonicalName(suffix));
    const fallback = nonempty(environment, legacyName(suffix));
    if (primary !== undefined && fallback !== undefined && primary !== fallback) value(suffix, environment);
    if (primary !== undefined && fallback === undefined) environment[legacyName(suffix)] = primary;
  }
  return environment;
}

export function exportBoth(
  values: Record<string, unknown>,
  options: { base?: Environment; scrubIdentity?: boolean } = {},
): Environment {
  const result: Environment = { ...(options.base ?? {}) };
  if (options.scrubIdentity) {
    for (const suffix of [...IDENTITY_SUFFIXES, ...TRANSIENT_IDENTITY_SUFFIXES]) {
      delete result[canonicalName(suffix)];
      delete result[legacyName(suffix)];
    }
  }
  for (const [suffix, raw] of Object.entries(values)) {
    if (raw === null || raw === undefined) continue;
    const rendered = String(raw);
    result[canonicalName(suffix)] = rendered;
    result[legacyName(suffix)] = rendered;
  }
  return result;
}

export function withProcessBoundary<T>(values: Record<string, unknown>, callback: () => T): T {
  const names = [
    ...[...IDENTITY_SUFFIXES, ...TRANSIENT_IDENTITY_SUFFIXES, "HOME", "SKILL_DIR"].flatMap((suffix) => [
      canonicalName(suffix),
      legacyName(suffix),
    ]),
  ];
  const before = Object.fromEntries(names.map((name) => [name, process.env[name]]));
  try {
    for (const name of names) delete process.env[name];
    Object.assign(process.env, exportBoth(values));
    return callback();
  } finally {
    for (const name of names) delete process.env[name];
    for (const [name, prior] of Object.entries(before)) {
      if (prior !== undefined) process.env[name] = prior;
    }
  }
}

function expandHome(path: string): string {
  if (path === "~") return homedir();
  if (path.startsWith("~/")) return `${homedir()}/${path.slice(2)}`;
  return path;
}

export function runtimeHome(environment: Environment = process.env): string {
  const configured = value("HOME", environment);
  return resolve(expandHome(configured ?? `${homedir()}/${RUNTIME_HOME_DIRECTORY}`));
}

export function runtimeSqlitePath(environment: Environment = process.env): string {
  return `${runtimeHome(environment)}/${RUNTIME_SQLITE_FILENAME}`;
}

export const runtimeRoot = runtimeHome;
export const dbPath = runtimeSqlitePath;
